File: demos/mini-distributed-store/store/ring.py
"""Consistent hash ring with virtual nodes (see kb/07).

Keys and nodes are hashed onto a 2^32 ring. A key is owned by the first vnode
clockwise; replicas are the next distinct *physical* nodes clockwise. Virtual
nodes smooth load distribution and minimize key movement when membership
changes -- the same idea CRUSH/Dynamo/Cassandra use.
"""

import bisect
import hashlib
from typing import Dict, List


def _hash(key: str) -> int:
    return int(hashlib.md5(key.encode()).hexdigest(), 16) & 0xFFFFFFFF


class HashRing:
    def __init__(self, vnodes: int = 128):
        self.vnodes = vnodes
        self._ring: Dict[int, str] = {}     # position -> physical node
        self._sorted: List[int] = []        # sorted positions
        self.nodes: set = set()

    def add_node(self, node: str):
        if node in self.nodes:
            return
        self.nodes.add(node)
        for i in range(self.vnodes):
            pos = _hash(f"{node}#{i}")
            self._ring[pos] = node
            bisect.insort(self._sorted, pos)

    def remove_node(self, node: str):
        if node not in self.nodes:
            return
        self.nodes.discard(node)
        for i in range(self.vnodes):
            pos = _hash(f"{node}#{i}")
            if pos in self._ring:
                del self._ring[pos]
                idx = bisect.bisect_left(self._sorted, pos)
                if idx < len(self._sorted) and self._sorted[idx] == pos:
                    self._sorted.pop(idx)

    def replicas(self, key: str, n: int) -> List[str]:
        """First n distinct physical nodes clockwise from hash(key)."""
        if not self._sorted:
            return []
        pos = _hash(key)
        idx = bisect.bisect(self._sorted, pos) % len(self._sorted)
        out: List[str] = []
        seen = set()
        for _ in range(len(self._sorted)):
            node = self._ring[self._sorted[idx]]
            if node not in seen:
                seen.add(node)
                out.append(node)
                if len(out) == n:
                    break
            idx = (idx + 1) % len(self._sorted)
        return out

    def owner(self, key: str) -> str:
        r = self.replicas(key, 1)
        return r[0] if r else ""

File: web/src/sim/ring.js
// Client-side consistent hash ring (see kb/07 & demos/mini-distributed-store).

// 32-bit string hash (xmur3-ish) -> position on a 2^32 ring.
function hash32(str) {
  let h = 1779033703 ^ str.length
  for (let i = 0; i < str.length; i++) {
    h = Math.imul(h ^ str.charCodeAt(i), 3432918353)
    h = (h << 13) | (h >>> 19)
  }
  h = Math.imul(h ^ (h >>> 16), 2246822507)
  h = Math.imul(h ^ (h >>> 13), 3266489909)
  return (h ^= h >>> 16) >>> 0
}

export class HashRing {
  constructor(vnodes = 64) {
    this.vnodes = vnodes
    this.ring = new Map()       // pos -> node
    this.sorted = []            // sorted positions
    this.nodes = new Set()
  }
  _insert(pos, node) {
    this.ring.set(pos, node)
    let lo = 0, hi = this.sorted.length
    while (lo < hi) { const m = (lo + hi) >> 1; if (this.sorted[m] < pos) lo = m + 1; else hi = m }
    this.sorted.splice(lo, 0, pos)
  }
  addNode(node) {
    if (this.nodes.has(node)) return
    this.nodes.add(node)
    for (let i = 0; i < this.vnodes; i++) this._insert(hash32(`${node}#${i}`), node)
  }
  removeNode(node) {
    if (!this.nodes.has(node)) return
    this.nodes.delete(node)
    for (let i = 0; i < this.vnodes; i++) {
      const pos = hash32(`${node}#${i}`)
      this.ring.delete(pos)
      const idx = this.sorted.indexOf(pos)
      if (idx >= 0) this.sorted.splice(idx, 1)
    }
  }
  replicas(key, n) {
    if (!this.sorted.length) return []
    const pos = hash32(key)
    let lo = 0, hi = this.sorted.length
    while (lo < hi) { const m = (lo + hi) >> 1; if (this.sorted[m] <= pos) lo = m + 1; else hi = m }
    let idx = lo % this.sorted.length
    const out = [], seen = new Set()
    for (let c = 0; c < this.sorted.length && out.length < n; c++) {
      const node = this.ring.get(this.sorted[idx])
      if (!seen.has(node)) { seen.add(node); out.push(node) }
      idx = (idx + 1) % this.sorted.length
    }
    return out
  }
  // vnode positions normalized to [0,1) for drawing
  vnodePositions() {
    return this.sorted.map(p => ({ pos: p / 4294967296, node: this.ring.get(p) }))
  }
}

export { hash32 }
